pp_netlib/network.py

```diff
@@ -137,38 +141,51 @@ def construct(self, network_data, weights = None): #, previous_network = None, a
 
-        ## keeping this section here for now; might be useful in add_to_network method
-        # ########################
-        # ####  PREV NETWORK  ####
-        # ########################
-        # if previous_network is not None:
-        #     prev_edges = []
-        #     if weights is not None:
-        #         extra_sources, extra_targets, extra_weights = process_previous_network(previous_network = previous_network, adding_qq_dists = adding_qq_dists, old_ids = old_ids, previous_pkl = previous_pkl, vertex_labels = vertex_labels, weights = (weights is not None), use_gpu = use_gpu)
-        #         for (src, dest, weight) in zip(extra_sources, extra_targets, extra_weights):
-        #                 prev_edges.append((src, dest, weight))
 
-        #     else:
-        #         extra_sources, extra_targets = process_previous_network(previous_network = previous_network, adding_qq_dists = adding_qq_dists, old_ids = old_ids, previous_pkl = previous_pkl, vertex_labels = vertex_labels, weights = (weights is not None), use_gpu = use_gpu)
-        #         for (src, dest, weight) in zip(extra_sources, extra_targets):
-        #                 prev_edges.append((src, dest))
+    def prune(self, type_isolate = None):
 
+        if self.graph is None:
+            raise RuntimeError("Graph not constructed or loaded.")
 
-        #     self.graph.add_edge_list(prev_edges) ## add previous edge list to newly made graph
+        if self.backend == "GT":
+            reference_vertices = set()
+            components = self.gt.label_components(self.graph)[0].a
 
-        #     self.edges = [edge for edge in self.graph.edges()]
+            for component in set(components):
+                reference_indices = gt_clique_prune(component, self.graph, set(), components)
+                reference_vertices.update(reference_indices)
 
+            labels = list(self.graph.vp["id"][v] for v in self.graph.vertices())
+            self.graph = gt_get_ref_graph(self.graph, reference_vertices, labels, type_isolate, backend=self.backend)
+            num_nodes = self.graph.num_vertices()
+            num_edges = self.graph.num_edges()
 
-    def prune(self):
-        #prune_cliques() ###TODO populate this function call with arguments
-        return
+        if self.backend == "NX":
+            reference_vertices = alt_nx_get_clique_refs(self.graph, set())
+
+            num_refs = len(reference_vertices)
+            # print(f"reference_vertices = {reference_vertices}, num_refs = {num_refs}")
+            updated_refs = nx_get_connected_refs(self.graph, reference_vertices)
+
+            type_idx = [i[0] for i in list(self.graph.nodes(data="id")) if i[1] == type_isolate]
+            # print(f"\n\n\ntype_idx = {type_idx}\n\n\n")
+            updated_refs.add(type_idx[0])
+            # print(f"\n\nupdated_refs = {updated_refs}\n\n")
+            self.graph.remove_nodes_from([node for node in self.graph.nodes() if node not in updated_refs])
```

Comment:
Don't overwrite `self.graph`, as above

Reply:
fixed by first making a copy of `self.graph` then pruning that